torch/csrc/jit/codegen/cuda/scheduler/matmul.cpp

```diff
@@ -0,0 +1,218 @@
+  TORCH_INTERNAL_ASSERT(mma != nullptr);
+  auto ab = mma->inA()->as<TensorView>();
+  auto bb = mma->inB()->as<TensorView>();
+
+  // Get exact configurations from mma builder.
+  auto mma_options = mma_builder.build();
+
+  // Staging register for global memory load
+  auto ar = a->cacheAfter();
+  auto br = b->cacheAfter();
+
+  // TODO:
+  //  Significant build out needed here
+  //   for more flexibility and data type support.
+  // Shared memory
+  TensorView* acw = nullptr;
+  TensorView* bcw = nullptr;
+  // Shared memory read
+  TensorView* acr = nullptr;
+  TensorView* bcr = nullptr;
+
+  if (isVolta(mma_options.macro)) {
+    acw = ab->cacheAfter();
+    bcw = bb->cacheAfter();
+    // Cache again to be able to vectorize.
+    acw = acw->cacheAfter();
+    bcw = bcw->cacheAfter();
+
+    acr = acw->cacheAfter();
+    bcr = bcw->cacheAfter();
+  } else {
+    acw = ar->cacheAfter();
+    bcw = br->cacheAfter();
+    acr =
+        acw->cacheAfter(mma_builder.operand(MmaOptions::Operand::A).ldMatrix());
+    bcr =
+        bcw->cacheAfter(mma_builder.operand(MmaOptions::Operand::B).ldMatrix());
+  }
+
+  // Make a CTA tile
+  // ------------------------------------------------------------------
+  scheduler_utils::matmul_utils::canonicalizeMmaTvOrdering(cc);
+  // [... M,N,K]
+  scheduler_utils::matmul_utils::makeTile(cc, gemm_tile.cta_tile.toVector());
+
+  // [Mo, No, Ko, Mi, Ni, Ki]
+  // Propagate tiling globally
+  scheduler_utils::matmul_utils::transformPropagateToAllFrom(cc, -1);
+
+  // Schedule warp tile
+  scheduler_utils::matmul_utils::scheduleWarpTileWithReduction(cc, gemm_tile);
+
+  // Propagate warp tile to main loop and epilog/output tvs
+  scheduler_utils::matmul_utils::transformPropagateWithin(
+      cc, -1, {acw, bcw}, {c});
+
+  // Schedule prolog:
+  //   TODO: this section goes to a separate matmul util,
+  //   and needs more configurability.
+  // ------------------------------------------------------------------
+  scheduler_utils::matmul_utils::orderTiledConcreteIdAsRoot(acw);
+  // [... M, K]
+  acw->merge(-2);
+  scheduler_utils::matmul_utils::scheduleContiguousVectorLoad(
+      acw, gemm_tile, 8, false);
+
+  // [... N, K]
+  scheduler_utils::matmul_utils::orderTiledConcreteIdAsRoot(bcw);
+  bcw->merge(-2);
+  scheduler_utils::matmul_utils::scheduleContiguousVectorLoad(
+      bcw, gemm_tile, 8, false);
+
+  // Propagate prolog tensors
+  //  propagate up the DAG, and propagate parallel type.
+  scheduler_utils::matmul_utils::transformPropagateWithin(
+      acw, -1, {a}, {}, true);
+  scheduler_utils::matmul_utils::transformPropagateWithin(
+      bcw, -1, {b}, {}, true);
+
+  // Set computeAt, setup the loop nesting structure on the kernel.
+  //   TODO: this section goes to a separate matmul util,
+  //   and needs more configurability.
+  // ------------------------------------------------------------------
+  // CTA tile:
+  a->computeAt(c, 2);
+  b->computeAt(c, 2);
```

Comment:
Maybe `cc->computeAt(c, 2)` instead of the above two lines would be less confusion here.